enhancements/0002-20221026-experimental-rest-transport-layer.md

```diff
@@ -0,0 +1,156 @@
+
+The motivation for this FED is, therefore, to make it easier for users to integrate, use, and deploy Flower in environments that have traditionally not been well supported.
+
+### Goals
+
+- *Increase infrastructure compatibility.* default configurations of common cloud infrastructure stacks.
+  - *Be compatible with caching infrastructure.*
+  - *Support CORS.*
+- *Increase client platform compatibility.* Flower clients should be able to run on platforms that are not well-supported by gRPC.
+- *Support proxy servers.* Enable client/server connections to go through proxy servers that may or may not alter the request, for example, to conceal the client's IP address from the server.
+- *Keep client identity optional.* Clients should be able to remain anonymous, client identity should be an optional component that users of the framework can opt for depending on the needs of their workload.
+- *Prevent clients from poisoning the model.*
+- *Limit server-side knowledge about individual clients.*
+
+### Non-Goals
+
+- *End-to-end latency.* The existing bidirectional stream between client and server offers stellar end-to-end latency (i.e., from scheduling a task for a client until that client starts executing the task locally). Most workloads do not require such low levels of latency.
+- *API stability.* The proposed REST API is not intended to be final.
+- *Client authentication.* There should be a forward-compatible path for adding client authentication, but client authentication does not need to be implemented in the first version.
+- *Compatibility with `start_server`.*
+- *Build a "RESTful" API.* The goal is to have a REST-inspired API, but not to build a fully idiomatic RESTful API.
+
+## Proposal
+
+Introduce a request/response REST transport layer. The REST transport layer would consist of a REST API server (integrated with the Flower server) and a REST client.
+
+## Design Details
+
+### Definition(s)
+
+Task: User-defined specification of what a client has to do, with the expectation that the client returns a result.
+
+### Protocol
+
+General idea:
+- Client always initiates conversation
+- Server responds in one of two ways:
+  - Reconnect at a later time.
+  - Execute a task and send the result.
+- Server has a (mild) form of control over the conversation by suggesting to the client when it should connect again.
+
+Example:
+- Client connects to the server: "I'm available"
+- Server doesn't have any tasks: "reconnect in 5min"
+- Client sleeps for 5min, then connects again
+- Server say: here's a TASK and a TOKEN, work on it and send me your RESULT
```

Comment:
Would it be possible (not sure how this ties in with the multi-tenancy update) to send down a list of available tasks instead of a single one? We were thinking of having some locally enforced task-specific eligibility criteria (i.e. does client have more than X samples for task A?), so not all tasks might be _runnable_ by a client that has advertised its availability. The current version with the single TASK shipped down can be a special case of the list of TASKs

Reply:
leaving the client to pick the task may add to the inefficiency of the protocol, but a simple priority/pacing system for the task might help with that

Reply:
Actually, the implementation will return a list of tasks instead of a single task. We need to update the document.

Reply:
True, the wording should have been more precise, I've updated the doc